Clarify concurrency requirements for trace and metrics (#720)
open-telemetry/opentelemetry-specification: specification/concurrency.md

The following are general recommendations of concurrent call safety of
specific subsets of the API.

## Trace API

> **TracerProvider** - all methods are safe to be called concurrently.

Reviewer: Also shutdown?

Author: It should be thread safe (no memory corruption or other problems related to thread safety), behavior is a different story, probably should be a no-op the second time, but I believe it should be safe and user should not implement a synchronization mechanism to ensure called only once.

**Span** - All methods of Span are safe to be called concurrently.

> **Event** - Events are immutable and is safe to be used concurrently. Lazy

Reviewer 2: ```suggestion
**Event** - Events are immutable and safe to be used concurrently. Lazy
```

> initialized events must be implemented to be safe to be called concurrently.

Reviewer: Really? Is there a case where the same event would be initialized more than once? Maybe multiple exporters?

Author: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#add-events

If an interface is defined by the language calls to that API should be thread safe. Also this is similar with existing Link

Reviewer: Then I suggest

```suggestion
initialized events must be thread safe. This is the responsibility of the implementer of these events.
```

Author: Done.

> initialized eventss must be implemented to be safe to be called concurrently.

Reviewer 3: ```suggestion
initialized events must be implemented to be safe to be called concurrently.
```

**Link** - Links are immutable and is safe to be used concurrently. Lazy
initialized links must be implemented to be safe to be called concurrently.

## Metrics

**MeterProvider** - all methods are safe to be called concurrently.

**Meter** - all methods are safe to be called concurrently.

**Instrument** - All methods of any Instrument are safe to be called concurrently.

> **Bound** - All methods of any Bound are safe to be called concurrently.

Reviewer 4: ```suggestion
**Bound Instrument** - All methods of any Bound Instrument are safe to be called concurrently.
```
